### ISSUE TITLE
WriteTag not triggered when on ignore

### ISSUE THREAD

**Issue by msteiger** [checkstyle/checkstyle#952]
We're currently trying to get rid of all @author tags in JavaDoc comments.
I'm trying to create a rule based on [WriteTag](http://checkstyle.sourceforge.net/config_javadoc.html#WriteTag) to flag the remaining ones, but it doesn't do anything.

    <module name="WriteTag">
      <property name="severity" value="ignore"/>
      <property name="tag" value="@author"/>
      <property name="tagSeverity" value="warning"/>
      <message key="javadoc.writeTag" value="No {0} tags should be used."/>
    </module>

The problem is the `severity` property. Setting it to `ignore` seems to make CheckStyle skip the entire thing, which it shouldn't.

Setting it to `info` instead flags the tags just fine, but also reports when they are missing.
Note that this snippet is almost identical to the example that is provided in the docs.

**Comment by romani** [comment]
after investigation I can confirm that this Check is completely broken and whole idea is damanged by design.

It has to be deprecated and removed from Checkstyle at all or reimplemented completely and be based on new JavaDocAst tree, rather then RegExp. New implementation should not have two severity properties but could solve tasks that this issue describe.

Workaround - please use any Check from http://checkstyle.sourceforge.net/config_regexp.html

**Comment by sdavids** [comment]
Could you please at least update the doc in the meantime, thx.

**Comment by romani** [comment]
Please send a PR, if you know how to explain what is going on in this Check, whole documentation in sources of this repo.

**Comment by ImmortalRabbit** [comment]
> Please send a PR, if you know how to explain what is going on in this Check, whole documentation in sources of this repo.

I can try to update documentation. 

I can put a note that says that usage of "<property name="severity" value="ignore"/>" doesn't work and can put a link to this issue in the Description part of WriteTag.

**Comment by rnveach** [comment]
@ImmortalRabbit Issue is not that property doesn't work. You will have to look at logic of check and try to document it's logic.

**Comment by ImmortalRabbit** [comment]
> @ImmortalRabbit Issue is not that property doesn't work. You will have to look at logic of check and try to document it's logic.

```
    @Test
    public void testSeverity() throws Exception {
        final DefaultConfiguration checkConfig = createModuleConfig(WriteTagCheck.class);
        checkConfig.addAttribute("tag", "@author");
        checkConfig.addAttribute("tagFormat", "\\S");
        checkConfig.addAttribute("severity", "ignore");
        final String[] expected = {
            "10: " + getCheckMessage(MSG_WRITE_TAG, "@author", "Daniel Grenner"),
        };
        verify(checkConfig, getPath("InputWriteTag.java"), expected);
    }
```

This test is passing. I think there is some errors with parsing from xml. Any other ideas what it could be?

@romani @rnveach

**Comment by ImmortalRabbit** [comment]
> @ImmortalRabbit Issue is not that property doesn't work. You will have to look at logic of check and try to document it's logic.

I have done some research. And I didn't 100% logic of check but I have found that. Some WriteTagCheckTests doesn't work properly because those tests are passing. 

```

It was discussed that we could remove WriteTag: 
    public void testSeverity() throws Exception {
        final DefaultConfiguration checkConfig = createModuleConfig(WriteTagCheck.class);
        checkConfig.addAttribute("tag", "@author");
        checkConfig.addAttribute("tagFormat", "\\S");
        checkConfig.addAttribute("severity", "ignore");
        final String[] expected = {
           "10: " + getCheckMessage(MSG_WRITE_TAG, "@author", "Daniel Grenner"),
        };
        verify(checkConfig, getPath("InputWriteTag.java"), expected);
}
    public void testSeverity() throws Exception {
        final DefaultConfiguration checkConfig = createModuleConfig(WriteTagCheck.class);
        checkConfig.addAttribute("tag", "@author");
        checkConfig.addAttribute("tagFormat", "\\S");
        final String[] expected = CommonUtil.EMPTY_STRING_ARRAY;
        verify(checkConfig, getPath("InputWriteTag.java"), expected);
}
```

**Comment by romani** [comment]
@gianmarcoschifone , please share how to archive `We're currently trying to get rid of all @author tags in JavaDoc comments.` by this Check , full CLI output.
It would be awesome to update checkstyle-checks.xml to do this validation. And add this to Use case in xdoc, as this is very frequent ask to kill this ancient tag. In 2026, there are hundreds of authors of files, and nobody own files, crowd ownership is around for many years. Old school code should remove old authors, that highly likely already not around.

Please share your opinion on removal of `tagSeverity` it might be preexisting before all Checks got their common property `severity`. By looking at doc https://checkstyle.org/checks/javadoc/writetag.html I don't understand how it works.

We can create separate issue on removal

**Comment by gianmarcoschifone** [comment]
@romani to raise error when author tag occurs:
```
roman@roman-precision-5510:~/tmp$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <!-- without this filter, missed @author will be violated -->
  <module name="SuppressionSingleFilter">
    <property name="checks" value="WriteTag"/>
    <property name="message"
        value="^Type Javadoc comment is missing @author tag\.$"/>
  </module>

  <module name="TreeWalker">
    <module name="WriteTag">
      <property name="tag" value="@author"/>
      <!-- not required if filter is present -->
      <!-- <property name="tagSeverity" value="error"/> -->
      <message key="javadoc.writeTag" value="No {0} tags should be used."/>
    </module>
  </module>
</module>

$ cat Test.java
/**
 * @author Me
 */
public class Test { }

/**
 * no tags.
 */
class Test2 { }

$ java -jar checkstyle-13.8.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:2: No @author tags should be used. [WriteTag]
Audit done.
Checkstyle ends with 1 errors.
```

WriteTag naturally enforces "**to write**" a specific tag, not "**to not write**". To achieve the desired behavior of forbidding one tag, users should use Filter to suppress natural violation of WriteTag.

---------------------

`tagSeverity` becomes useful, when users do not want to use Filters, BUT in this case they must use for Check:
```
 <property name="severity" value="ignore"/>
 <property name="tagSeverity" value="error"/>
```
AND
Ant validation runs with `executeIgnoredModules="true"`  https://checkstyle.org/anttask.html#Parameters
or CLI `-E` (`--executeIgnoredModules`)  https://checkstyle.org/cmdline.html#E


example:
```
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">

  <module name="TreeWalker">
    <module name="WriteTag">
      <property name="severity" value="ignore"/>
      <property name="tag" value="@author"/>
      <property name="tagSeverity" value="error"/>
      <message key="javadoc.writeTag" value="No {0} tags should be used."/>
    </module>
  </module>
</module>

$ cat Test.java
/**
 * @author Me
 */
public class Test { }

/**
 * no tags.
 */
class Test2 { }

$ java -jar checkstyle-13.8.0-all.jar -c config.xml Test.java
Starting audit...
Audit done.

# ATTENTION to "-E"
$ java -jar checkstyle-13.8.0-all.jar -E -c config.xml Test.java
Starting audit...
[ERROR] Test.java:2: No @author tags should be used. [WriteTag]
Audit done.
Checkstyle ends with 1 errors.
```


[javadoc.writeTag](https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%2Fcom%2Fpuppycrawl%2Ftools%2Fcheckstyle%2Fchecks%2Fjavadoc+path%3A**%2Fmessages*.properties+repo%3Acheckstyle%2Fcheckstyle+%22javadoc.writeTag%22) is printed when the configured tag is present, and [tagSeverity](https://checkstyle.sourceforge.io/checks/javadoc/writetag.html#tagSeverity) controls the severity of that message, which is set to `info` as default. Removing `tagSeverity` means that [javadoc.writeTag](https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%2Fcom%2Fpuppycrawl%2Ftools%2Fcheckstyle%2Fchecks%2Fjavadoc+path%3A**%2Fmessages*.properties+repo%3Acheckstyle%2Fcheckstyle+%22javadoc.writeTag%22) will be controlled by `severity`.

**Comment by romani** [comment]
@gianmarcoschifone , why SuppressionSingleFilteris used ?

**Comment by romani** [comment]
> It would be awesome to update checkstyle-checks.xml to do this validation.

please send PR to update it, lets keep valuable usage over our code.

**Comment by gianmarcoschifone** [comment]
> [@gianmarcoschifone](https://github.com/gianmarcoschifone) , why SuppressionSingleFilteris used ?

@romani SuppressionSingleFilter was used only to make the standalone CLI example work without `-E`.

`WriteTag` reports two different cases:

1. `javadoc.writeTag` when the configured tag is present.
   This is what we want for `@author`.

2. `type.missingTag` when the configured tag is absent.
   This is not desired for this use case, because absence of `@author` is valid.

So the filter suppresses only the second case, while keeping the `javadoc.writeTag` violation.

But for `config/checkstyle-checks.xml`, I think we can avoid `SuppressionSingleFilter`, because Checkstyle's own Ant validation runs with `executeIgnoredModules="true"`. In that case we can use:

```xml
<module name="WriteTag">
  <property name="tag" value="@author"/>
  <property name="severity" value="ignore"/>
  <property name="tagSeverity" value="error"/>
  <message key="javadoc.writeTag" value="No {0} tags should be used."/>
</module>

**Comment by romani** [comment]
@gianmarcoschifone , I updated your comment to have all details shared on how to make it work.

**Comment by romani** [comment]
@gianmarcoschifone , please help to update examples https://checkstyle.org/checks/javadoc/writetag.html#Examples
to mention extra details in for tagSeverity, and add Usecase (author prevention, with filter) to show user how to make it.

usage of `-E` is very technical and should not be used by users until they know what they are doing. We cannot recommend this approach.

**Comment by romani** [comment]
New issues are created:
https://github.com/checkstyle/checkstyle/issues/21079
https://github.com/checkstyle/checkstyle/issues/21080

**Comment by romani** [comment]
documentation update is merged.
